docs/locks.rst

```diff
@@ -108,3 +108,66 @@ multithreaded app.)*
         About to set the event
         Not waiting this time
         Done
+
+   Semaphore
+   ---------
+   .. autoclass:: Semaphore
+    :members:
+
+    Semaphores limit access to a shared resource. To allow access for two
+    workers at a time:
+
+    .. testsetup:: semaphore
+
+       from tornado import gen
+
+       # Ensure reliable doctest output.
+       waits = [0.1, 0.2, 0.1]
+
+       @gen.coroutine
+       def use_some_resource():
+           yield gen.sleep(waits.pop())
+
+    .. testcode:: semaphore
+
+        sem = locks.Semaphore(2)
+
+        @gen.coroutine
+        def worker(worker_id):
+            yield sem.acquire()
+            try:
+                print("Worker %d is working" % worker_id)
+                yield use_some_resource()
+            finally:
+                print("Worker %d is done" % worker_id)
+                sem.release()
+
+        @gen.coroutine
+        def runner():
+            # Join all workers.
+            yield [worker(i) for i in range(3)]
+
+        io_loop.run_sync(runner)
+
+    .. testoutput:: semaphore
+
+        Worker 0 is working
+        Worker 1 is working
+        Worker 0 is done
+        Worker 2 is working
+        Worker 1 is done
+        Worker 2 is done
```

Comment:
These last two lines are non-deterministic; I just saw this in a CI run:

```
File "locks.rst", line 263, in semaphore
Failed example:
    sem = locks.Semaphore(2)
    @gen.coroutine
    def worker(worker_id):
        yield sem.acquire()
        try:
            print("Worker %d is working" % worker_id)
            yield use_some_resource()
        finally:
            print("Worker %d is done" % worker_id)
            sem.release()
    @gen.coroutine
    def runner():
        # Join all workers.
        yield [worker(i) for i in range(3)]
    io_loop.run_sync(runner)
Expected:
    Worker 0 is working
    Worker 1 is working
    Worker 0 is done
    Worker 2 is working
    Worker 1 is done
    Worker 2 is done
Got:
    Worker 0 is working
    Worker 1 is working
    Worker 0 is done
    Worker 2 is working
    Worker 2 is done
    Worker 1 is done
1 items passed all tests:
   2 tests in default
**********************************************************************
1 items had failures:
   1 of   1 in semaphore
3 tests in 2 items.
2 passed and 1 failed.
***Test Failed*** 1 failures.
```

Looks like a 100ms difference in sleep time isn't enough for reliable ordering on travis; we'll need to rig things up even more deterministically.